src/util/KinBodyParser.cpp

```diff
@@ -0,0 +1,532 @@
+  if (kinBodyDoc.Parse(kinBodyString.c_str()) != tinyxml2::XML_SUCCESS)
+  {
+    kinBodyDoc.PrintError();
+    return nullptr;
+  }
+
+  return readKinBody(kinBodyDoc, baseUri, nullOrRetriever);
+}
+
+//==============================================================================
+dart::dynamics::SkeletonPtr readKinbody(
+  const dart::common::Uri& kinBodyFileUri,
+  const dart::common::ResourceRetrieverPtr& nullOrRetriever)
+{
+  // Parse XML file
+  tinyxml2::XMLDocument kinBodyDoc;
+  try
+  {
+    dart::utils::openXMLFile(kinBodyDoc, kinBodyFileUri, nullOrRetriever);
+  }
+  catch(std::exception const& e)
+  {
+    dterr << "[KinBodyParser] LoadFile '" << kinBodyFileUri.toString()
+          << "' Fails: " << e.what() << "\n";
+    return nullptr;
+  }
+
+  return readKinBody(kinBodyDoc, kinBodyFileUri, nullOrRetriever);
+}
+
+namespace {
+
+//==============================================================================
+dart::dynamics::SkeletonPtr readKinBody(
+  tinyxml2::XMLDocument& kinBodyDoc,
+  const dart::common::Uri& baseUri,
+  const dart::common::ResourceRetrieverPtr& nullOrRetriever)
+{
+  const auto retriever = getRetriever(nullOrRetriever);
+
+  // Transform all the element names to lower cases because the KinBody spec is
+  // case-insensitive. We use lower cases to read the XML elements.
+  transformElementNamesToLowerCases(kinBodyDoc);
+
+  auto kinBodyEle = kinBodyDoc.FirstChildElement("kinbody");
+  if (!kinBodyEle)
+  {
+    dterr << "[KinBodyParser] KinBody document does not contain <KinBody> as "
+          << "the root element.\n";
+    return nullptr;
+  }
+
+  return readKinBody(kinBodyEle, baseUri, retriever);
+}
+
+//==============================================================================
+dart::dynamics::SkeletonPtr readKinBody(
+    tinyxml2::XMLElement* kinBodyEle,
+    const dart::common::Uri& baseUri,
+    const dart::common::ResourceRetrieverPtr& retriever)
+{
+  assert(kinBodyEle != nullptr);
+  // kinBodyEle shouldn't be nullptr at all, which always should be
+  // gauranteed because this function is only used internally to be so. If we
+  // find the nullptr case, then we should fix the caller.
+
+  auto skeleton = dart::dynamics::Skeleton::create();
+
+  // Read name attribute.
+  //
+  // If name attribute doesn't exist, getAttributeString() returns empty string.
+  // Empty name for skeleton is allowed.
+  auto name = dart::utils::getAttributeString(kinBodyEle, "name");
+  if (name.empty())
+  {
+    dtwarn << "[KinBodyParser] The <KinBody> in '" << baseUri.toString()
+           << "' doesn't have name attribute or has empty name.\n";
+  }
+  skeleton->setName(name);
+
+  // Get Body node
+  auto bodyEle = kinBodyEle->FirstChildElement("body");
+  if (bodyEle == nullptr)
+  {
+    dterr << "[KinBodyParser] KinBody document '" << baseUri.toString()
+          << "' does not contain <Body> element "
+          << "under <KinBody> element.\n";
+    return nullptr;
+  }
```

Comment:
Nit: I'd like to raise an error if there is more than one `body` element as well. Otherwise, we may end up silently loading only the first `BodyNode` in the file!